// braintree/lib/braintree.js
// Generated by CoffeeScript 1.6.1
var BraintreeGateway, Config, CreditCard, CreditCardDefaults, CreditCardNumbers, CreditCardVerification, Environment, MerchantAccount, MerchantAccountTest, Nonces, PaymentInstrumentTypes, Subscription, TestingGateway, Transaction, TransactionAmounts, VenmoSdk, WebhookNotification, connect, errorTypes;

Config = require("./braintree/config").Config;

Environment = require("./braintree/environment").Environment;

BraintreeGateway = require("./braintree/braintree_gateway").BraintreeGateway;

errorTypes = require("./braintree/error_types");

Transaction = require("./braintree/transaction").Transaction;

CreditCard = require("./braintree/credit_card").CreditCard;

CreditCardVerification = require("./braintree/credit_card_verification").CreditCardVerification;

Subscription = require("./braintree/subscription").Subscription;

MerchantAccount = require("./braintree/merchant_account").MerchantAccount;

PaymentInstrumentTypes = require("./braintree/payment_instrument_types").PaymentInstrumentTypes;

WebhookNotification = require("./braintree/webhook_notification").WebhookNotification;

TestingGateway = require("./braintree/testing_gateway").TestingGateway;

CreditCardDefaults = require("./braintree/test/credit_card_defaults").CreditCardDefaults;

CreditCardNumbers = require("./braintree/test/credit_card_numbers").CreditCardNumbers;

MerchantAccountTest = require("./braintree/test/merchant_account").MerchantAccountTest;

Nonces = require("./braintree/test/nonces").Nonces;

TransactionAmounts = require("./braintree/test/transaction_amounts").TransactionAmounts;

VenmoSdk = require("./braintree/test/venmo_sdk").VenmoSdk;

connect = function(config) {
  return new BraintreeGateway(new Config(config));
};

exports.connect = connect;

exports.version = '1.29.0';

exports.Environment = Environment;

exports.errorTypes = errorTypes;

exports.Transaction = Transaction;

exports.CreditCard = CreditCard;

exports.CreditCardVerification = CreditCardVerification;

exports.Subscription = Subscription;

exports.MerchantAccount = MerchantAccount;

exports.WebhookNotification = WebhookNotification;

exports.TestingGateway = TestingGateway;

exports.Test = {
  CreditCardDefaults: CreditCardDefaults,
  CreditCardNumbers: CreditCardNumbers,
  MerchantAccountTest: MerchantAccountTest,
  Nonces: Nonces,
  TransactionAmounts: TransactionAmounts
};
